apps/src/javalab/JavalabFileManagement.js

```diff
@@ -27,24 +25,31 @@ function renameProjectFile(filename, newFilename) {
   );
 }
 
-function onSave() {
-  header.showProjectSaving();
-  const filename = getStore().getState().javalab.filename;
-  const editorText = getStore().getState().javalab.editorText;
+function onSave(success, failure) {
   // TODO: enable multi-file
-  filesApi.putFile(
-    filename,
-    editorText,
-    () => {
-      // at this point project.filesVersionId will be updated and we can save
-      // sources as well.
-      project.save();
-    },
-    () => {
-      console.warn(`Javalab: error file ${filename} not saved`);
-      project.showSaveError_();
+  const filesChanged = getStore().getState().javalab.filesChanged;
+  // only save file if file content has changed.
+  if (filesChanged) {
+    const filename = getStore().getState().javalab.filename;
+    const editorText = getStore().getState().javalab.editorText;
+    filesApi.putFile(
+      filename,
+      editorText,
+      /* success */
+      (xhr, filesVersionId) => {
+        // reset files changed to false on success
+        getStore().dispatch(setFilesChanged(false));
+        if (success) {
```

Comment:
This code path would run if `success` was any truthy value. `typeof success === "function"` might be closer to what you're intending.

Reply:
(similarly below)